### PR TITLE
Add flexibility to directory management in tomcat::instance

### PR DESCRIPTION
**Pull request by esalberg**
Some Tomcat installations have non-standard directory layouts under $catalina_base.

Allow disabling $catalina_base sub-directories entirely via $manage_dirs
Allow configuration of which $catalina_base sub-directories are managed vi $dir_list (array)
Allow configuration of the mode for $catalina_base sub-directories.
Retain all defaults (dir_list and mode set in tomcat::params)

### CODE REVIEW

**Comment by esalberg** [comment]
Please let me know if you need acceptance tests for this - I haven't written them before, but I can give it a try.

**Comment by esalberg** [comment]
I don't think the Travis errors are related to my code change, but please let me know if they are.

**Comment by esalberg** [comment]
Something isn't working right with dir_list - I'm doing some troubleshooting and hoping I won't have to pull the PR.

**Comment by esalberg** [comment]
Rebased for init.pp updates.

File: manifests/init.pp [hunk 1]
```diff
@@ -27,18 +27,32 @@
...
+#   Defaults to tomcat::params::dir_list.
+#
+# [*dir_mode*]
+#   Mode to use for managed sub-directories under $catalina_base. Defaults to '2770'.
 class tomcat (
   $catalina_home       = $::tomcat::params::catalina_home,
   $user                = $::tomcat::params::user,
   $group               = $::tomcat::params::group,
+  Array $dir_list            = $::tomcat::params::dir_list,
+  String $dir_mode           = $::tomcat::params::dir_mode,
```
Comment (hunner) [inline manifests/init.pp]: I'd rather not add more parameters to the base class; see https://github.com/puppetlabs/puppetlabs-tomcat/pull/145 for a detailed reasoning.

**Comment by hunner** [comment]
Thanks for submitting this and #218. These are definitely areas that the module is not very flexible at the moment, but I'd really appreciate describing the reasoning behind supplying the ability to give a list of directories to copy or disabling the copying altogether.

Could you elaborate on the use cases for this, and perhaps link to related documentation? I've tried to be as "mainstream" as possible with the functionality of this module and these open that up quite a bit...

**Comment by esalberg** [comment]
We were given an app-supplied Tomcat rpm, which our app team insisted on using because it was "supported" by the vendor.  It actually ended up being compatible with the module, except for having a strange directory listing.  I yum installed the rpm to see what it did, then backfilled the info in the module so that the module could manage the service, connectors, etc.  I didn't actually have docs from the vendor (sorry about that).

App: Symphony
Package: tomcat-8.0.41-25.x86_64.rpm (RHEL7)
catalina_home: /data/tomcat
catalina_base: /opt/tomcat
I ended up with a dir_list of ['bin','lib','native','webapps','webappsbot']
I also had to make a symlink of ${catalina_home}/conf -> ${catalina_home}/config

I'm fine with just being able to disable copying the default directories, if you'd prefer not to open the option of setting those directories to anything you might want.  That seems rather less open to non-mainstream customization, and I can always manage the directories in my profile if needed.

**Comment by hunner** [comment]
(Thanks for keeping these updated with rebases!)

Okay that makes sense. Another PR came in (#210) to also make the copy_to_home list more flexible as well (and is related to your #218 change) so `manage_copy_from_home`, `copy_from_home_list`,  `manage_base_dirs`, and `base_dirs_list` may be a generally helpful set of parameters.

Also, I've started moving away from using init.pp to set defaults as it can create more "spaghetti code" than having the profile define set its own defaults. Eg:
```ruby
#spaghetti
class profile::tomcat_base {
  include profile::base
  class { 'tomcat':
    dir_mode => '2750',
  }
}
define profile::tomcat_instance($dir_mode = undef) {
  include profile::tomcat_base
  tomcat::install { ... }
  tomcat::instance { $name:
    dir_mode => $dir_mode,
  }
}
```
vs.
```ruby
#not spaghetti
define profile::tomcat_base($dir_mode = '2750') {
  include profile::base
  tomcat::install { ... }
  tomcat::instance { $name:
    dir_mode => $dir_mode,
  }
}
```
I wrote more about this in #145 and would be interested in your thoughts on this direction?

**Comment by esalberg** [comment]
You mentioned this previously and I completely forgot and moved the params around in init.pp during rebase.

I agree that if the parameters are only used in one manifest, then why not specify them there directly.  Let me see about updating my code to match, and if I run into any concerns I'll update again.

**Comment by esalberg** [comment]
Code has been updated to move the parameters to tomcat::instance.  I also rearranged the tomcat::instance section of the README.md file.

I did notice that the instance.pp file doesn't have parameters listed in Puppet Strings format, nor are the booleans typed.  I didn't want to add that in as part of this PR specifically, but do you need a PR for that cleanup?

**Comment by hunner** [comment]
Sorry for the delay on this and #218 ; I've been trying to focus on only what is needed to get the 2.0.0 out the door (as it is long overdue) and these two PRs are backwards-compatible so can go in the 2.1.0 release or so.

**Comment by esalberg** [comment]
Rebased against master

**Comment by david22swan** [comment]
Resolved conflicts from merging the other pr

**Pull request merged**